File: project/BBWT.Client/src/app/features/bb-formio/dto/form-revision.ts
import {FormDefinitionForRequestMinDTO, FormIODefinition} from "./form-definition";

export interface FormRevisionDTO {
    id?: string;
    id_original?: number;

    dateCreated?: Date;
    note?: string;
    json?: string;
    mufCapable?: boolean;
    mobileFriendly?: boolean;

    // Foreign keys and navigational properties.
    creatorId?: string;
    formDefinitionId?: number | string;
    formDefinition?: FormIODefinition;
}

export interface FormRevisionMinDTO {
    id: string;
    id_original: number;

    dateCreated: Date;
    mufCapable?: boolean;
    majorVersion: number;
    minorVersion: number;
    json: string;
    mobileFriendly: boolean;
}

export class InitialFormRevisionRequest {
    mobileFriendly?: boolean;
    json: string;
    mufCapable?: boolean;
}

export class NewFormRevisionRequest {
    mobileFriendly: boolean;
    creatorId?: string;
    note?: string
    json: string;
    mufCapable?: boolean;

    // Foreign keys and navigational properties.
    formDefinitionId: string | number;
}

export class UpdateFormRevisionRequest {
    mobileFriendly: boolean;
    note?: string
    json: string;
    increaseMinorVersion: boolean;
    saveAsMajorVersion: boolean;
    mufCapable?: boolean;

    // Foreign keys and navigational properties.
    creatorId?: string;
    formDefinitionName: string;
}

export class FormRevisionForRequestMinDTO {
    id: string;

    id_original: number;
    creatorName?: string;

    // Foreign keys and navigational properties.
    formDefinitionId: string;
    formDefinition: FormDefinitionForRequestMinDTO;
}

export class FormRevisionForFormDataPageDTO extends FormRevisionForRequestMinDTO {
}

File: modules/BBWM.Formio/Models/FormRevision.cs
using AutoMapper;
using BBWM.Core.Data;
using BBWM.Core.Membership.Model;
using BBWM.FormIO.DTO;

namespace BBWM.FormIO.Models
{
    public class FormRevision : IEntity
    {
        public int Id { get; set; }

        public int MajorVersion { get; set; }

        public int MinorVersion { get; set; }

        public DateTimeOffset DateCreated { get; set; }

        /// <summary>
        /// Indicates if a form can be filled on mobile devices
        /// </summary>
        public bool MobileFriendly { get; set; } = false;

        public string? CreatorId { get; set; }
        public User? Creator { get; set; }

        public string? Note { get; set; }

        public int? FormDefinitionId { get; set; }
        public FormDefinition? FormDefinition { get; set; }

        public string? Json { get; set; }

        public bool? MUFCapable { get; set; }

        public ICollection<FormSurvey> Surveys { get; set; } = new List<FormSurvey>();

        public static void RegisterMap(IMapperConfigurationExpression c)
        {
            c.CreateMap<FormRevision, FormRevisionDTO>()
                .ForMember(dest => dest.CreatorName, opt => opt.MapFrom(x => x.Creator.UserName))
                .ReverseMap()
                .ForMember(dest => dest.Creator, opt => opt.Ignore());

            c.CreateMap<FormRevision, FormRevisionMinDTO>()
                .ReverseMap();

            c.CreateMap<FormRevision, InitialFormRevisionRequestDTO>()
                .ReverseMap()
                .ForMember(dest => dest.Creator, opt => opt.Ignore());

            c.CreateMap<FormRevision, NewFormRevisionRequestDTO>()
                .ReverseMap()
                .ForMember(dest => dest.Creator, opt => opt.Ignore());

            c.CreateMap<FormRevision, FormRevisionForRequestMinDTO>()
                .ForMember(dest => dest.CreatorName, opt => opt.MapFrom(x => x.Creator!.UserName))
                .ReverseMap();

            c.CreateMap<FormRevision, FormRevisionForFormDataPageDTO>()
                .ForMember(dest => dest.CreatorName, opt => opt.MapFrom(x => x.Creator!.UserName))
                .ReverseMap();
        }
    }
}